pkg/operator/ceph/object/user/controller.go

```diff
@@ -124,6 +126,61 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 		return err
 	}
 
+	// watch secrets referenced by CephObjectStoreUser.spec.keys
+	const (
+		secretNameField = "spec.keys.secretNames"
+	)
+
+	err = mgr.GetFieldIndexer().IndexField(context.TODO(), &cephv1.CephObjectStoreUser{TypeMeta: controllerTypeMeta}, secretNameField, func(obj client.Object) []string {
+		var secretNames []string
+		for _, k := range obj.(*cephv1.CephObjectStoreUser).Spec.Keys {
+			if k.AccessKeyRef != nil && k.AccessKeyRef.Name != "" {
+				secretNames = append(secretNames, k.AccessKeyRef.Name)
+			}
+			if k.SecretKeyRef != nil && k.SecretKeyRef.Name != "" {
+				secretNames = append(secretNames, k.SecretKeyRef.Name)
+			}
+		}
+		return secretNames
+	})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(
+		source.Kind(
+			mgr.GetCache(),
+			&corev1.Secret{},
+			handler.TypedEnqueueRequestsFromMapFunc(
+				func(ctx context.Context, secret *corev1.Secret) []reconcile.Request {
+					referencingUsers := &cephv1.CephObjectStoreUserList{}
+					err := r.(*ReconcileObjectStoreUser).client.List(ctx, referencingUsers, &client.ListOptions{
+						FieldSelector: fields.OneTermEqualSelector(secretNameField, secret.GetName()),
+						Namespace:     secret.GetNamespace(),
+					})
+					if err != nil {
+						return []reconcile.Request{}
+					}
+
+					requests := make([]reconcile.Request, len(referencingUsers.Items))
+					for i, item := range referencingUsers.Items {
+						requests[i] = reconcile.Request{
+							NamespacedName: types.NamespacedName{
+								Name:      item.GetName(),
+								Namespace: item.GetNamespace(),
+							},
+						}
+					}
+					return requests
```

Comment:
This return doesn't include the possibility of returning an error, so I don't think controller-runtime has the ability to requeue the event that kicked this process off later when it thinks that it handled it as needed. I think this is a risk for missing events if the `client.List()` fails.

Reply:
I believe this is true for all watch handlers.  There is no mechanism to requeue the attempt and it won't be triggered again until there is another predicate allowed event.  I'm not sure if there is a guarantee that k8s will deliver all events either.

If the cluster is unable to list resources, there's a good chance that any other action attempted by the operator is going to fail as well.  I think the best that can be done here is to prove that the logic is correct with an integration test.

Reply:
Logging of errors has been added to watch and field index setup.